kickstarter/android-oss
# NT-1958: UX – Post a reply

# 📲 What

UX – Post a reply

# 🤔 Why

Time to handle how we post a reply within a thread!
- New reply is appended to the existing list of replies
- View scrolls to the bottom of the replies, showing the newly posted reply

# 🛠 How

- Refactored ``` CommentsViewHolderViewModel ``` to handle replies
```
PostCommentData(
                        commentableId = it.first,
                        body = it.second.body(),
                        clientMutationId = null,
                        parent = it.second?.parentId()?.let { id -> it.second.toBuilder().id(id).build() }
                    )

```

- Added post observables to ``` ThreadViewModel ```
```
 this.insertNewReplyToList
                .distinctUntilChanged()
                .withLatestFrom(this.currentUser.loggedInUser()) {
                    comment, user ->
                    Pair(comment, user)
                }
                .withLatestFrom(commentData) {
                    reply, parent ->
                    Pair(reply, parent)
                }
                .map {
                    Pair(it.second, buildReplyBody(Pair(Pair(it.second, it.first.second), it.first.first)))
                }
                .map {
                    CommentCardData.builder()
                        .comment(it.second)
                        .project(it.first.project)
                        .commentableId(it.first.commentableId)
                        .commentCardState(CommentCardStatus.TRYING_TO_POST.commentCardStatus)
                        .build()
                }
                .withLatestFrom(this.onCommentReplies) { reply, list ->
                    list.toMutableList().apply {
                        add(reply)
                    }.toList()
                }.compose(bindToLifecycle())
                .subscribe {
                    onCommentReplies.onNext(it)
                    scrollToBottom.onNext(it.size - 1)
                }
```

# 👀 See

https://user-images.githubusercontent.com/63934292/123618002-b21fe880-d7ff-11eb-9be4-812c0a04953d.mp4


| Before 🐛 | After 🦋 |
| --- | --- |
|  |  |

# 📋 QA

- Click on ``` reply ``` or ``` view replies ``` button in the comment section of a backed project
- Type a reply and post. You see new reply appended to the replies and scrolled to bottom

# Story 📖

https://kickstarter.atlassian.net/browse/NT-1958
